### PR TITLE
Nenad/150 abci spec fixes

### LINKED ISSUE #150: Fixing ABCI spec inconsistencies.
ABCI spec has inconsistencies regarding when PrepareProposal and ProcessProposal are called. 
Specifically, in the "Overview and basic concepts" section, it is stated that both are called only if a process hasn't locked on a value, while in the "Methods" section, the conditions are different:
 
1) For PrepareProposal, a valid value is consulted instead (valid value != nil doesn't necessarily mean that locked value is != nil). 

2) For ProcessProposal, it is written that it is called whenever a process should send a Prevote message (notice that this can happen even when the process has locked value != nil, [[arXiv paper line 30]](https://github.com/tendermint/spec/blob/master/spec/consensus/consensus-paper/consensus.tex#L83)). 

Spec should be fixed to match what is written in "Overview and basic concepts". 
In addition, a file explaining why a process doesn't need to call PrepareProposal and ProcessProposal if it has a locked value != nil should be added.

### PR DESCRIPTION
**Pull request by nenadmilosevic95**
---
Closes #150 

#### PR checklist

- [x] Removed mention of locked value from the text.
- [x] Modified grammar in file 'abci++_tmint_expected_behavior.md' to account for rounds where the proposer does not call `PrepareProposal` and `ProcessProposal`.
- [x] Added file with additional scenarios ('spec/abci/abci++_extra.md'). 
- [x] File 'spec/abci/abci++_extra.md' integrated and polished. 
- [ ] Renaming done (Tendermint -> CometBFT).

### CODE REVIEW

**Review by sergio-mena — APPROVED** [review]
Thanks Nenad for this writeup. I really helps the app developer have a better idea that runs that are odd but totally possible, and so, the app's logic should be ready to handle them.

File: spec/abci/abci++_extra.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    
+1. On a correct node, `PrepareProposal` may be called multiple times and for different blocks ([**Scenario 1**](#scenario-1)). 
+1. On a correct node, `ProcessProposal` may be called multiple times and for different blocks ([**Scenario 2**](#scenario-2)).
+1. On a correct node, `PrepareProposal` and `ProcessProposal` for block $X$ may not be called ([**Scenario 3**](#scenario-3)).
+1. On a correct node, `PrepareProposal` and `ProcessProposal` may not be called at all ([**Scenario 4**](#scenario-4)).
+
+
+## Basic information
+
+Each scenario is presented from the perspective of a process $p$. More precisely, we show what happens in each round's $step$ of the [Tendermint algorithm](https://arxiv.org/pdf/1807.04938.pdf). The legend is below:
```
Comment (josef-widder) [inline spec/abci/abci++_extra.md:44]:
```suggestion
Each scenario is presented from the perspective of a process $p$. More precisely, we show what happens in each round's $step$ of the [Tendermint algorithm](https://arxiv.org/pdf/1807.04938.pdf). While in practice the consensus algorithm works with respect to voting power of the validators, in this document refer to number of processes (e.g., $n$, $f+1$, $2f+1$) for simplicity. The legend is below: 
```

File: spec/abci/abci++_extra.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+1. **Propose:** Describes what happens while $step_p = propose$.
+1. **Prevote:** Describes what happens while $step_p = prevote$.
+1. **Precommit:** Describes what happens while $step_p = precommit$.
+
+## Scenario 1
+
+### Round 0:
+    
+1. **Propose:** Proposer of this round is a Byzantine process, and it chooses not to send the proposal message. Therefore, $p$'s $timeoutPropose$ expires, it sends $Prevote$ for $nil$, and it does not call `ProcessProposal`. All correct processes do the same. 
+1. **Prevote:** $p$ eventually receives $2f+1$ $Prevote$ messages for $nil$ and starts $timeoutPrevote$. When $timeoutPrevote$ expires it sends $Precommit$ for $nil$.
```
[the remark this answers is not in the document]
Comment (josef-widder) [reply spec/abci/abci++_extra.md:59]: we added a sentence in the intro

File: spec/abci/abci++_methods.md [hunk 3]
```diff
@@ -550,7 +550,7 @@ When a node _p_ enters Tendermint consensus round _r_, height _h_, in which _q_
 4. Upon reception of Proposal message, along with all the block parts, for round _r_, height _h_
    from _q_, _p_'s Tendermint follows the validators' algorithm to check whether it should prevote for the
    proposed block, or `nil`.
-5. If the validators' algorithm indicates Tendermint should prevote for the proposed block:
+5. If the validators' algorithm indicates Tendermint should prevote for the proposed block and _p_'s _validValue_ is `nil`:
```
Comment (josef-widder) [inline spec/abci/abci++_methods.md:553]:
```suggestion
5. If the validator's Tendermint consensus instance indicates that $p$ should prevote non-nil:
```

File: spec/abci/abci++_basic_concepts.md [hunk 4]
```diff
@@ -78,8 +78,8 @@ call sequences of these methods.
   proposer to perform application-dependent work in a block before proposing it.
   This enables, for instance, batch optimizations to a block, which has been empirically
   demonstrated to be a key component for improved performance. Method `PrepareProposal` is called
-  every time CometBFT is about to broadcast a Proposal message, but no previous proposal has
-  been locked at consensus level. CometBFT gathers outstanding transactions from the
+  every time CometBFT is about to broadcast a Proposal message and CometBFT's _validValue_ is `nil`.
```
Comment (cason) [inline spec/abci/abci++_basic_concepts.md:81]:
```suggestion
  every time CometBFT is about to broadcast a Proposal message and _validValue_ is `nil`. 
```

File: spec/abci/abci++_basic_concepts.md [hunk 5]
```diff
@@ -78,8 +78,8 @@ call sequences of these methods.
   proposer to perform application-dependent work in a block before proposing it.
   This enables, for instance, batch optimizations to a block, which has been empirically
   demonstrated to be a key component for improved performance. Method `PrepareProposal` is called
-  every time CometBFT is about to broadcast a Proposal message, but no previous proposal has
-  been locked at consensus level. CometBFT gathers outstanding transactions from the
+  every time CometBFT is about to broadcast a Proposal message and CometBFT's _validValue_ is `nil`. 
+  CometBFT gathers outstanding transactions from the
```
Comment (cason) [inline spec/abci/abci++_basic_concepts.md:82]: I new line/paragraph here would be helpful.

File: spec/abci/abci++_basic_concepts.md [hunk 6]
```diff
@@ -90,8 +90,9 @@ call sequences of these methods.
 - [**ProcessProposal:**](./abci++_methods.md#processproposal) It allows a validator to
   perform application-dependent work in a proposed block. This enables features such as immediate
   block execution, and allows the Application to reject invalid blocks.
-  CometBFT calls it when it receives a proposal and the consensus algorithm has not locked on a
-  value. The Application cannot modify the proposal at this point but can reject it if it is
+
+  CometBFT calls it when it receives a proposal and CometBFT's _validValue_ is `nil`.
```
Comment (cason) [inline spec/abci/abci++_basic_concepts.md:94]: validValue or lockedValue?

Comment (cason) [reply spec/abci/abci++_basic_concepts.md:94]: I would merge this sentence to the previous paragraph.

**Comment by cason** [comment]
I wonder whether we could use this PR to rename `abci++_tmint_expected_behavior.md` to `abci++_comet_expected_behavior.md` or something like that.

There are no references to this file name in the repository, except for `spec/abci/README.md` which is also part of this PR.

File: spec/abci/abci++_basic_concepts.md (hunk 6 above)
Comment (cason) [reply spec/abci/abci++_basic_concepts.md:94]:
```suggestion
  CometBFT calls it when it receives a proposal and _validValue_ is `nil`. 
```

File: spec/abci/abci++_extra.md [hunk 7]
```diff
@@ -41,7 +41,7 @@ The unknown number of rounds we can have when following the Tendermint algorithm
 
 ## Basic information
 
-Each scenario is presented from the perspective of a process $p$. More precisely, we show what happens in each round's $step$ of the [Tendermint algorithm](https://arxiv.org/pdf/1807.04938.pdf). The legend is below: 
+Each scenario is presented from the perspective of a process $p$. More precisely, we show what happens in each round's $step$ of the [Tendermint algorithm](https://arxiv.org/pdf/1807.04938.pdf). While in practice the consensus algorithm works with respect to voting power of the validators, in this document refer to number of processes (e.g., $n$, $f+1$, $2f+1$) for simplicity. The legend is below:
```
Comment (lasarojc) [inline spec/abci/abci++_extra.md]:
```suggestion
Each scenario is presented from the perspective of a process $p$. More precisely, we show what happens in each round's $step$ of the [Tendermint algorithm](https://arxiv.org/pdf/1807.04938.pdf). While in practice the consensus algorithm works with respect to voting power of the validators, in this document we refer to number of processes (e.g., $n$, $f+1$, $2f+1$) for simplicity. The legend is below: 
```

**Comment by sergio-mena** [comment]
> I wonder whether we could use this PR to rename abci++_tmint_expected_behavior.md to abci++_comet_expected_behavior.md or something like that.

I support that (I had missed it in my review)

File: spec/abci/abci++_example_scenarios.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+1. Call `PrepareProposal` and/or `ProcessProposal` for block $X$. 
+1. Call `PrepareProposal` and/or `ProcessProposal` for block $Y \neq X$.
+1. Does not call `PrepareProposal` and/or `ProcessProposal`.
+
+In the rounds when it is the proposer, Tendermint's `PrepareProposal` call is always followed by the `ProcessProposal` call. The reason is that the process always delivers the proposal to itself, which triggers the `ProcessProposal` call. 
+
+As the number of rounds the Tendermint consensus algorithm needs to decide in a given run is a priori unknown, the application needs to account for any number of rounds, where each round can exhibit any of these three behaviours. Recall that the application is unaware of the internals of consensus and thus of the rounds. 
+
+# Possible scenarios
+The unknown number of rounds we can have when following the Tendermint algorithm yields a vast number of scenarios we can expect. Listing them all is impossible. However, here we give several of them and draw the main conclusions. Specifically, we will show that before `decide` and `commit` for block $X$ are called:
```
[the remark this answers is not in the document]
Comment (nenadmilosevic95) [reply spec/abci/abci++_example_scenarios.md]: Fixed in new version.

[the remark this answers is not in the document]
Comment (nenadmilosevic95) [reply spec/abci/abci++_example_scenarios.md]: Fixed in the new version.

File: spec/abci/abci++_basic_concepts.md (hunk 5 above)
Comment (nenadmilosevic95) [reply spec/abci/abci++_basic_concepts.md:82]: Fixed in the new version.

File: spec/abci/abci++_extra.md (hunk 7 above)
Comment (nenadmilosevic95) [reply spec/abci/abci++_extra.md]: Fixed in the new version.

**Review by cason — APPROVED** [review]

**Pull request merged**